### PR TITLE
kokkos: initial implementation

### PR DESCRIPTION
**Pull request by tom91136**
Will begin adding scripts on the other repo while waiting for merge.

### CODE REVIEW

File: kokkos/src/bude.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#include <iomanip>
+
+#ifdef KOKKOS_IN_TREE
+	#include <Kokkos_Core.hpp>
+#else
+	#include <kokkos/Kokkos_Core.hpp>
+#endif
+
+#ifndef DEFAULT_WGSIZE
+#define DEFAULT_WGSIZE 64
```
Comment (andreipoe) [inline kokkos/src/bude.h:14]: Make the default size 128. This is a good choice on both CXL and Rome.

File: kokkos/src/vec-pose-inner.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+		} while (++ip < natpro); // loop over protein atoms
+	} while (++il < natlig); // loop over ligand atoms
+
+	// Write results
+	for (size_t l = 0; l < WG_SIZE; l++) {
+		// Write result
+		etotals[group * WG_SIZE + l] = etot[l] * 0.5f;
+	}
+
+
```
Comment (andreipoe) [inline kokkos/src/vec-pose-inner.cpp:157]: Remove this spacing.

**Review by andreipoe — APPROVED** [review]
This works fine on the MACS platform. I'm approving it now so we can merge it if you want, but I'm going to try the Arm platforms too.

File: kokkos/src/bude.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#include <string>
+#include <iomanip>
+
+#ifdef KOKKOS_IN_TREE
+	#include <Kokkos_Core.hpp>
+#else
+	#include <kokkos/Kokkos_Core.hpp>
+#endif
+
+#ifndef DEFAULT_WGSIZE
```
[the remark this answers is not in the document]
Comment (tom91136) [reply kokkos/src/bude.h:13]: It shows up in the help message, so:
```

File: kokkos/src/bude.h (hunk 1 above)
Comment (tom91136) [reply kokkos/src/bude.h:14]: Done

File: kokkos/src/vec-pose-inner.cpp (hunk 2 above)
Comment (tom91136) [reply kokkos/src/vec-pose-inner.cpp:157]: Done

File: kokkos/src/bude.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (tom91136) [reply kokkos/src/bude.h:13]: Oops, `WG_SIZE` isn't a runtime arg anymore on kokkos. We should eventually consolidate all the different terms. Maybe keep only two, so something like:
 * `UNROLL_ITER` for the static vector/loop unroll thing. Should be available in all impl.
 * `WG_SIZE` only exists for implementations with `nd_range`, such as CL, CUDA, and SYCL
And no more `NUM_TD_PER_THREAD` or `ppWI`

File: kokkos/src/vec-pose-inner.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+#define NPNPDIST 5.5f
+#define NPPDIST  1.0f
+
+constexpr const auto FloatMax = std::numeric_limits<float>::max();
+
+void fasten_main(
+		size_t group,
+		size_t ntypes, size_t nposes,
+		size_t natlig, size_t natpro,
+		Kokkos::View<Atom *> protein_molecule,
```
[the remark this answers is not in the document]
Comment (tom91136) [reply kokkos/src/vec-pose-inner.cpp]: Done, I've also made the component type const, so `const Kokkos::View<const Atom *> &protein_molecule ... `

File: kokkos/src/bude.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+#ifndef WG_SIZE
+#define WG_SIZE DEFAULT_WGSIZE
+#endif
+
+#define DEFAULT_ITERS  8
+#define DEFAULT_NPOSES 65536
+#define REF_NPOSES     65536
+
+#define DATA_DIR          "../data"
```
[the remark this answers is not in the document]
Comment (tom91136) [reply kokkos/src/bude.h]: Done

File: kokkos/src/bude.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (tom91136) [reply kokkos/src/bude.h]: I'm opening an issue for this so we can fix it later

**Pull request merged**